Dualstack

jose-ledesma opened the pull request:
<!-- HTML Comments can be left in place or removed. -->
<!-- Please see our contributing guide at https://github.com/rancher/k3s/blob/master/CONTRIBUTING.md for guidance on opening pull requests -->

#### Proposed Changes ####

<!-- Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. -->

#### Types of Changes ####

<!-- What types of changes does your code introduce to K3s? Bugfix, New Feature, Breaking Change, etc -->

#### Verification ####

<!-- How can the changes be verified? Please provide whatever additional information necessary to help verify the proposed changes. -->

#### Linked Issues ####

<!-- Link any related issues, pull-requests, or commit hashes that are relevant to this pull request. If you are opening a PR without a corresponding issue please consider creating one first, at https://github.com/rancher/k3s/issues . A functional example will greatly help QA with verifying/reproducing a bug or testing new features. -->

#### Further Comments ####

<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

erikwilson committed:
Add diagnostics collection scripts

Signed-off-by: Erik Wilson <Erik.E.Wilson@gmail.com>

Commit:
try to make dualstack work

c3y1huang committed:
Fix typo

Signed-off-by: Chin-Ya Huang <chin-ya.huang@suse.com>

Commit:
multiple external test

Commit:
multiple node ips

Commit:
move after to before

Commit:
move addresses to annotations

Commit:
hostname to annotation + server command before

Commit:
add service-cluster-ip-range arg to controller-manager

Commit:
Add codespell CI test and fix codespell error (#2740)

* Add codespell CI test
* Fix codespell error

erikwilson committed:
Bootstrap node password with local file

Signed-off-by: Erik Wilson <Erik.E.Wilson@gmail.com>

c3y1huang committed:
Traefik v2 integration

K3s upgrade via watch over file change of static file and manifest
and triggers helm-controller for change. It seems reasonable to
only allow upgrade traefik v1->v2 when there is no existing custom
traefik HelmChartConfig in the cluster to avoid any
incompatibility.

Here also separate the CRDs and put them into a different chart
to support CRD upgrade.

Signed-off-by: Chin-Ya Huang <chin-ya.huang@suse.com>

erikwilson committed:
Fix spelling mistake

Signed-off-by: Erik Wilson <Erik.E.Wilson@gmail.com>

erikwilson committed:
Return http.StatusOK instead of 0

Signed-off-by: Erik Wilson <Erik.E.Wilson@gmail.com>

Commit:
Fix typo

Signed-off-by: JenTing Hsiao <jenting.hsiao@suse.com>

brandond committed:
Add v1.20 channel and mark as stable

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

erikwilson committed:
Merge pull request #2526 from erikwilson/diagnostics-script

Add diagnostics collection

Commit:
change etcd dir permission if it exists

Signed-off-by: MonzElmasry <menna.elmasry@rancher.com>

Commit:
Fix normal user with --rootless or --disable-agent have no permission to start

Signed-off-by: JenTing Hsiao <jenting.hsiao@suse.com>

MonzElmasry committed:
Merge pull request #2784 from MonzElmasry/etcd-dir

change etcd dir permission if it exists

c3y1huang committed:
Allow download traefik static file and rename

Allow writing static files regardless of the version.

Signed-off-by: Chin-Ya Huang <chin-ya.huang@suse.com>

sonicaj committed:
Do not validate snapshotter argument if docker is enabled

Problem:
While using ZFS on debian and K3s with docker, I am unable to get k3s working as the snapshotter value is being validated and the validation fails.

Solution:
We should not validate snapshotter value if we are using docker as it's a no-op in that case.

Signed-off-by: Waqar Ahmed <waqarahmedjoyia@live.com>

erikwilson committed:
Merge pull request #2744 from erikwilson/rke2-node-password-bootstrap

Bootstrap node password with local file

brandond committed:
Update helm-controller to v0.8.3 (#2793)

* Update helm-controller to v0.8.3
* Update klipper-helm image

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

rybit commented on pkg/daemons/control/server.go:
```diff
@@ -134,16 +134,29 @@ func controllerManager(cfg *config.Control, runtime *config.ControlRuntime) erro
-		"cluster-signing-cert-file":        runtime.ClientCA,
-		"cluster-signing-key-file":         runtime.ClientCAKey,
+		"cluster-cidr": func() string {
+			ranges := make([]string, len(cfg.ClusterIPRanges))
```
I'd pull this into a helper function. You do it in a few places

rybit reviewed (commented):
Generally looks good to me. I'd make sure to add some context on how you're adding Annotations vs Labels, that code is a bit confusing to follow.

rybit commented on pkg/agent/config/config.go:
```diff
@@ -81,7 +81,7 @@ func getNodeNamedCrt(nodeName, nodeIP, nodePasswordFile string) HTTPRequester {
 		}
 		req.Header.Set(version.Program+"-Node-Password", nodePassword)
-		req.Header.Set(version.Program+"-Node-IP", nodeIP)
+		req.Header.Set(version.Program+"-Node-IP", nodeIPs[0])
```
So there is a lot of fallout in signatures to make them start taking `[]string`, then we throw that all away by taking the 0th element. 

An alternative that might be a bit cleaner is to actually not change the signature (b/c to me it hides the fact that we're only using the 0 position) and push that responsibility higher in the call chain. So somewhere in the `get` method that calls `getServingCert` and `getNodeNamedHostFile`

brandond committed:
Update Kubernetes to v1.20.2-k3s1

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

Commit:
mark k3s v1.20.02 as stable

Signed-off-by: MonzElmasry <menna.elmasry@rancher.com>

brandond committed:
Fix typo in field tag

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Only container-runtime-endpoint wants RuntimeSocket path as URI

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

davidnuzik committed:
Merge pull request #2821 from MonzElmasry/mark_1.20.2

mark k3s v1.20.2 as stable

briandowns committed:
Add ability to perform an etcd on-demand snapshot via cli (#2819)

* add ability to perform an etcd on-demand snapshot via cli

brandond committed:
Update sonobuoy version for v1.20.0

Also drop the rancher-mirrored sonobuoy image since CI has a
pull-through image cache now.

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Improve sonobuoy status output handling

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

AkihiroSuda committed:
fix cgroup2 support

Fix issue 900

cgroup2 support was introduced in PR 2584, but got broken in f3de60ff3191dc3a0ae7ce8393e388700f48ed88

It was failing with "F1210 19:13:37.305388    4955 server.go:181] cannot set feature gate SupportPodPidsLimit to false, feature is locked to true"

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
rootless: remove rootful /run/{netns,containerd} symlinks

Since a recent commit, rootless mode was failing with the following errors:

```
E0122 22:59:47.615567      21 kuberuntime_manager.go:755] createPodSandbox for pod "helm-install-traefik-wf8lc_kube-system(9de0a1b2-e2a2-4ea5-8fb6-22c9272a182f)" failed: rpc error: code = Unknown desc = failed to create network namespace for sandbox "285ab835609387f82d304bac1fefa5fb2a6c49a542a9921995d0c35d33c683d5": failed to setup netns: open /var/run/netns/cni-c628a228-651e-e03e-d27d-bb5e87281846: permission denied
...
E0122 23:31:34.027814      21 pod_workers.go:191] Error syncing pod 1a77d21f-ff3d-4475-9749-224229ddc31a ("coredns-854c77959c-w4d7g_kube-system(1a77d21f-ff3d-4475-9749-224229ddc31a)"), skipping: failed to "CreatePodSandbox" for "coredns-854c77959c-w4d7g_kube-system(1a77d21f-ff3d-4475-9749-224229ddc31a)" with CreatePodSandboxError: "CreatePodSandbox for pod \"coredns-854c77959c-w4d7g_kube-system(1a77d21f-ff3d-4475-9749-224229ddc31a)\" failed: rpc error: code = Unknown desc = failed to create containerd task: io.containerd.runc.v2: create new shim socket: listen unix /run/containerd/s/8f0e40e11a69738407f1ebaf31ced3f08c29bb62022058813314fb004f93c422: bind: permission denied\n: exit status 1: unknown"
```

Remove symlinks to /run/{netns,containerd} so that rootless mode can create their own /run/{netns,containerd}.

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

brandond committed:
Update k3s-root to v0.7.3

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Replace k3s cloud provider wrangler controller with core node informer (#2843)

* Replace k3s cloud provider wrangler controller with core node informer

Upstream k8s has exposed an interface for cloud providers to access the
cloud controller manager's node cache and shared informer since
Kubernetes 1.9. This is used by all the other in-tree cloud providers;
we should use it too instead of running a dedicated wrangler controller.

Doing so also appears to fix an intermittent issue with the uninitialized
taint not getting cleared on nodes in CI.

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

yuriydzobak committed:
Add functionality to bind custom IP address for Etcd metrics endpoint (#2750)

* Add functionality to bind custom IP address for Etcd metrics endpoint

Signed-off-by: yuriydzobak <yurii.dzobak@lotusflare.com>

brandond committed:
Bump k3s-root to v0.8.0

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Spell check upstream code

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Update containerd to fix AppArmor on SLES

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Initial update of netpol and utils from upstream

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Add ServiceIPRange and ServiceNodePortRange to agent config

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Replace options.KubeRouterConfig with config.Node and remove metrics/waitgroup stuff

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

AkihiroSuda committed:
systemd unit: make EnvironmentFile optional

Previously, k3s.service was failing when the EnvironmentFile does not exist:
```
Feb 02 17:17:30 suda-ws01 systemd[1]: k3s.service: Failed to load environment files: No such file or directory
Feb 02 17:17:30 suda-ws01 systemd[1]: k3s.service: Failed to run 'start' task: No such file or directory
Feb 02 17:17:30 suda-ws01 systemd[1]: k3s.service: Failed with result 'resources'.
Feb 02 17:17:30 suda-ws01 systemd[1]: Failed to start Lightweight Kubernetes.
```

ref: https://unix.stackexchange.com/questions/404199/documentation-of-equals-minus-in-systemd-unit-files

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

brandond committed:
Update local-path-provisioner and helper busybox (#2885)

* Update local-path-provisioner and helper busybox

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Fix multiple issues with CLI wrapper data-dir handling

We also need to be more careful about setting the crictl.yaml path,
as it doesn't have kubectl's nice behavior of checking multiple
locations. It's not safe to assume that it's in the user's home data-dir
just because we're not running as root.

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Bump golang to 1.15.8

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Use zstd instead of gzip for embedded tarball

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

erikwilson committed:
Add opensuse 15 vagrant provision

brandond committed:
Allow joining clusters when the server CA is trusted by the OS CA bundle (#2743)

* Add tests to clientaccess/token
* Fix issues in clientaccess/token identified by tests
* Update tests to close coverage gaps
* Remove redundant check turned up by code coverage reports
* Add warnings if CA hash will not be validated

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Use appropriate response codes for authn/authz failures

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

erikwilson committed:
Merge pull request #2913 from erikwilson/vagrant-opensuse15

Add opensuse 15 vagrant provision

brandond committed:
Improve handling of comounted cpu,cpuacct controllers (#2911)

* Improve handling of comounted cpu,cpuacct controllers

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Collect IPs from all pods before deciding to use internal or external addresses (#2909)

* Collect IPs from all pods before deciding to use internal or external addresses

@Taloth correctly noted that the code that iterates over ServiceLB pods
to collect IP addresses was failing to add additional internal IPs once
the map contained ANY entry from a previous node. This may date back to
when ServiceLB used a Deployment instead of a DaemonSet, so there was
only ever a single pod.

The new behavior is to collect all internal and external IPs, and then
construct the address list of a single type - external if there are any,
otherwise internal.

https://github.com/k3s-io/k3s/issues/1652#issuecomment-774497788

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>
Co-authored-by: Brian Downs <brian.downs@gmail.com>

briandowns committed:
update usage text (#2926)

update to the --cluster-init usage flag to indicate it's for Etcd

erikwilson committed:
Remove Traefik v1 migration

erikwilson committed:
Update to Traefik 2.4.2 and combine manifests

brandond committed:
Add support for retagging images on load from tarball

Adds support for retagging images to appear to have been sourced from
one or more additional registries as they are imported from the tarball.
This is intended to support RKE2 use cases with system-default-registry
where the images need to appear to have been pulled from a registry
other than docker.io.

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

galal-hussein committed:
Add disable flags for control components (#2900)

* Add disable flags to control components

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

* golint

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

* more fixes

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

* fixes to disable flags

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

* Add comments to functions

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

* Fix joining problem

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

* more fixes

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

* golint

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

* fix ticker

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

* fix role labels

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

* more fixes

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

brandond committed:
Use HasSuffixI utility function

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Limit zstd decoder memory

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Update k3s-root to v0.8.1

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

galal-hussein committed:
update master to 1.20.4

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
use v1.20.4-k3s1 as stable

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

davidnuzik committed:
Merge pull request #2972 from galal-hussein/update_stable

mark v1.20.4-k3s1 as stable

erikwilson committed:
Apply suggestions from code review

Logging cleanup

Co-authored-by: Brad Davidson <brad@oatmail.org>

brandond committed:
Wait for apiserver to become healthy before starting agent controllers

It is possible that the apiserver may serve read requests but not allow
writes yet, in which case flannel will crash on startup when trying to
configure the subnet manager.

Fix this by waiting for the apiserver to become fully ready before
starting flannel and the network policy controller.

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Hide the airgap-extra-registry flag

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Add script to test server/agent version compatibility

We have had a couple issues with newer agents not working with old
servers or vice versa. Add a CI test to test variations on
uplevel/downlevel server/agent against latest, stable, and the previous
branch.

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Log clearer error on startup if NPC cannot be started

Servers should always be upgraded before agents, but generally this
isn't required because things are compatible between versions. In this
case we're OK with failing closed if the user upgrades out of order, but
we should give a clearer message about what steps are required to fix
the issue.

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

Commit:
Update .md files with url and email corrections

* BUILDING.md
* CODE_OF_CONDUCT.md
* CONTRIBUTING.md
* MAINTAINERS
* README.md

Signed-off-by: David Nuzik <david.nuzik@rancher.com>

Commit:
Update GITHUB_URL

Signed-off-by: David Nuzik <david.nuzik@rancher.com>

brandond committed:
Suppress test failure due to incompatible server

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

galal-hussein committed:
remove etcd member if disable etcd is passed

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
update dynamiclistener

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

erikwilson committed:
Merge pull request #2910 from erikwilson/traefik-v2

Traefik v2 integration

galal-hussein committed:
change error to warn when removing self from etcd members

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

galal-hussein committed:
Merge pull request #3003 from galal-hussein/fix_etcd_only_nodes

Fix etcd only nodes

briandowns committed:
Etcd Snapshot/Restore to/from S3 Compatible Backends (#2902)

* Add functionality for etcd snapshot/restore to and from S3 compatible backends.
* Update etcd restore functionality to extract and write certificates and configs from snapshot.

galal-hussein committed:
Mark disable components flags as experimental (#3018)

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

davidnuzik committed:
Merge pull request #3002 from davidnuzik/docs-housekeeping

Docs housekeeping

brandond committed:
Always use static ports for client load-balancers (#3026)

* Always use static ports for the load-balancers

This fixes an issue where RKE2 kube-proxy daemonset pods were failing to
communicate with the apiserver when RKE2 was restarted because the
load-balancer used a different port every time it started up.

This also changes the apiserver load-balancer port to be 1 below the
supervisor port instead of 1 above it. This makes the apiserver port
consistent at 6443 across servers and agents on RKE2.

Additional fixes below were required to successfully test and use this change
on etcd-only nodes.

* Actually add lb-server-port flag to CLI
* Fix nil pointer when starting server with --disable-etcd but no --server
* Don't try to use full URI as initial load-balancer endpoint
* Fix etcd load-balancer pool updates
* Update dynamiclistener to fix cert updates on etcd-only nodes
* Handle recursive initial server URL in load balancer
* Don't run the deploy controller on etcd-only nodes

brandond committed:
Handle loadbalancer port in TIME_WAIT

If the port wanted by the client load balancer is in TIME_WAIT, startup
will fail. Set SO_REUSEPORT so that it can be listened on again
immediately.

The configurable Listen call wants a context, so plumb that through as
well.

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

brandond committed:
Don't start up multiple apiserver load balancers

get() is called in a loop until client configuration is successfully
retrieved. Each iteration will try to configure the apiserver proxy,
which will in turn create a new load balancer. Skip creating a new
load balancer if we already have one.

Signed-off-by: Brad Davidson <brad.davidson@rancher.com>

Commit:
make vendor

Commit:
make generate

Commit:
rebase fixes

Commit:
more rebase fixes

Commit:
vendor

Commit:
regenerate
